core/src/muxing.rs

```diff
@@ -237,252 +166,3 @@ impl<T> StreamMuxerEvent<T> {
-    muxer: P,
-    substream: Option<<P::Target as StreamMuxer>::Substream>,
-    shutdown_state: ShutdownState,
-}
-
-enum ShutdownState {
-    Shutdown,
-    Flush,
-    Done,
-}
-
-impl<P> fmt::Debug for SubstreamRef<P>
-where
-    P: Deref,
-    P::Target: StreamMuxer,
-    <P::Target as StreamMuxer>::Substream: fmt::Debug,
-{
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
-        write!(f, "Substream({:?})", self.substream)
-    }
-}
-
-impl<P> Unpin for SubstreamRef<P>
-where
-    P: Deref,
-    P::Target: StreamMuxer,
-{
-}
-
-impl<P> AsyncRead for SubstreamRef<P>
-where
-    P: Deref,
-    P::Target: StreamMuxer,
-{
-    fn poll_read(
-        mut self: Pin<&mut Self>,
-        cx: &mut Context<'_>,
-        buf: &mut [u8],
-    ) -> Poll<Result<usize, io::Error>> {
-        // We use a `this` because the compiler isn't smart enough to allow mutably borrowing
-        // multiple different fields from the `Pin` at the same time.
-        let this = &mut *self;
-
-        let s = this.substream.as_mut().expect("substream was empty");
-        this.muxer.read_substream(cx, s, buf).map_err(|e| e.into())
-    }
-}
-
-impl<P> AsyncWrite for SubstreamRef<P>
-where
-    P: Deref,
-    P::Target: StreamMuxer,
-{
-    fn poll_write(
-        mut self: Pin<&mut Self>,
-        cx: &mut Context<'_>,
-        buf: &[u8],
-    ) -> Poll<Result<usize, io::Error>> {
-        // We use a `this` because the compiler isn't smart enough to allow mutably borrowing
-        // multiple different fields from the `Pin` at the same time.
-        let this = &mut *self;
-
-        let s = this.substream.as_mut().expect("substream was empty");
-        this.muxer.write_substream(cx, s, buf).map_err(|e| e.into())
-    }
-
-    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
-        // We use a `this` because the compiler isn't smart enough to allow mutably borrowing
-        // multiple different fields from the `Pin` at the same time.
-        let this = &mut *self;
-
-        let s = this.substream.as_mut().expect("substream was empty");
-        loop {
-            match this.shutdown_state {
-                ShutdownState::Shutdown => match this.muxer.shutdown_substream(cx, s) {
-                    Poll::Ready(Ok(())) => this.shutdown_state = ShutdownState::Flush,
-                    Poll::Ready(Err(err)) => return Poll::Ready(Err(err.into())),
-                    Poll::Pending => return Poll::Pending,
-                },
-                ShutdownState::Flush => match this.muxer.flush_substream(cx, s) {
-                    Poll::Ready(Ok(())) => this.shutdown_state = ShutdownState::Done,
-                    Poll::Ready(Err(err)) => return Poll::Ready(Err(err.into())),
-                    Poll::Pending => return Poll::Pending,
-                },
-                ShutdownState::Done => {
-                    return Poll::Ready(Ok(()));
-                }
-            }
-        }
```

Comment:
I think it is important to focus on this bit of deleted code here.

Previously, it would have been very likely that a substream created by _any_ `StreamMuxer` implementation would have gone through this logic because using the `StreamMuxer` interface without the help of `SubstreamRef` would be extremely cumbersome.

With this PR, substreams are forced to implement `AsyncRead` and `AsyncWrite` and thus, users can directly use them. Even if we put this piece of logic into `SubstreamBox`, it may be bypassed. The mplex implementation follows the pattern of "first close, then flush" anyway and is also separately tested.

My take here is: It is fine to delete this code as long as we document that all implementations of `StreamMuxer` must create substreams that flush after closing. If we ever get to do https://github.com/libp2p/rust-libp2p/issues/508, this could be part of that test suite. Perhaps we can extract such a test suite out of the mplex tests.

Reply:
> "first close, then flush" anyway and is also separately tested.

I am a bit confused here. My intuition would be: A call to `poll_close` first flushes the substream before closing it. Same as documented e.g. in Tokio's `AsyncWrite:

> Invocation of a shutdown implies an invocation of flush. Once this method returns Ready it implies that a flush successfully happened before the shutdown happened. That is, callers don’t need to call flush before calling shutdown. They can rely that by calling shutdown any pending buffered data will be written out.

https://docs.rs/tokio/1.6.0/tokio/io/trait.AsyncWrite.html#tymethod.poll_shutdown

Is this not the convention in libp2p? Based on the above it is not. Am I missing something?

Reply:
> Is this not the convention in libp2p? 

At least for mplex, no: https://github.com/libp2p/rust-libp2p/blob/ea487aebfe6eb672b05d2bec2d9d79bbd92450ba/muxers/mplex/src/io.rs#L527

I am not 100% sure why this code was added on this abstraction level. With your quote from above, I think this PR is actually making things more correct then because the now added `AsyncWrite` impl on `libp2p_mplex::Substream` implies flush on close.

Maybe @tomaka can provide some background why this was solved on a generic level previously?